contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java

```diff
@@ -0,0 +1,1046 @@
+      void recordConsumed(long offset, int size) {
+        consumedOffset = offset;
+
+        // this is always updated from single thread. probably not worth making it an AtomicDouble
+        if (avgRecordSize <= 0) {
+          avgRecordSize = size;
+        } else {
+          // initially, first record heavily contributes to average.
+          avgRecordSize += ((size - avgRecordSize) / movingAvgWindow);
+        }
+      }
+
+      synchronized void setLatestOffset(long latestOffset) {
+        this.latestOffset = latestOffset;
+      }
+
+      synchronized long approxBacklogInBytes() {
+        // Note that is an an estimate of uncompressed backlog.
+        // Messages on Kafka might be comressed.
+        if (latestOffset < 0 || consumedOffset < 0) {
+          return UnboundedReader.BACKLOG_UNKNOWN;
+        }
+        if (latestOffset <= consumedOffset || consumedOffset < 0) {
+          return 0;
+        }
+        return (long) ((latestOffset - consumedOffset - 1) * avgRecordSize);
+      }
+    }
+
+    public UnboundedKafkaReader(
+        UnboundedKafkaSource<K, V> source,
+        @Nullable KafkaCheckpointMark checkpointMark) {
+
+      this.source = source;
+      this.name = "Reader-" + source.id;
+
+      partitionStates = ImmutableList.copyOf(Lists.transform(source.assignedPartitions,
+          new Function<TopicPartition, PartitionState>() {
+            public PartitionState apply(TopicPartition tp) {
+              return new PartitionState(tp, -1L);
+            }
+        }));
+
+      if (checkpointMark != null) {
+        // a) verify that assigned and check-pointed partitions match exactly
+        // b) set consumed offsets
+
+        checkState(checkpointMark.getPartitions().size() == source.assignedPartitions.size(),
+            "checkPointMark and assignedPartitions should match");
+        // we could consider allowing a mismatch, though it is not expected in current Dataflow
+
+        for (int i = 0; i < source.assignedPartitions.size(); i++) {
+          PartitionMark ckptMark = checkpointMark.getPartitions().get(i);
+          TopicPartition assigned = source.assignedPartitions.get(i);
+
+          checkState(ckptMark.getTopicPartition().equals(assigned),
+              "checkpointed partition %s and assinged partition %s don't match at position %d",
+              ckptMark.getTopicPartition(), assigned, i);
+
+          partitionStates.get(i).consumedOffset = ckptMark.getOffset();
+        }
+      }
+    }
+
+    private void consumerPollLoop() {
+      // Read in a loop and enqueue the batch of records, if any, to availableRecordsQueue
+      while (!closed) {
+        try {
+          ConsumerRecords<byte[], byte[]> records = consumer.poll(1000);
+          if (!records.isEmpty()) {
+            availableRecordsQueue.put(records); // blocks until dequeued.
+          }
+        } catch (InterruptedException e) {
+          LOG.warn(this + " consumer thread is interrupted", e); // not expected
```

Comment:
http://stackoverflow.com/a/6374166/1715495

On Tue, Apr 5, 2016 at 10:38 AM, Raghu Angadi <notifications@github.com>
wrote:

> In
> contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java
> <https://github.com/GoogleCloudPlatform/DataflowJavaSDK/pull/121#discussion_r58581720>
> :
>
> > +
> > +          partitionStates.get(i).consumedOffset = ckptMark.getOffset();
> > +        }
> > +      }
> > +    }
> > +
> > +    private void consumerPollLoop() {
> > +      // Read in a loop and enqueue the batch of records, if any, to availableRecordsQueue
> > +      while (!closed) {
> > +        try {
> > +          ConsumerRecords<byte[], byte[]> records = consumer.poll(1000);
> > +          if (!records.isEmpty()) {
> > +            availableRecordsQueue.put(records); // blocks until dequeued.
> > +          }
> > +        } catch (InterruptedException e) {
> > +          LOG.warn(this + " consumer thread is interrupted", e); // not expected
>
> Done. Updated the all the LOG lines in the reader.
> JavaDoc was not clear about how exception is rendered with multiple
> objects as in LOG.warn("msg", o1, exception). Just tried it, it does the
> right thing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/DataflowJavaSDK/pull/121/files/3d560732adfb56d572740985bd4bc3df7900b1e3#r58581720>
>